File: src/main/java/com/github/pareronia/i18n_puzzles/Puzzle2025_08.java
package com.github.pareronia.i18n_puzzles;

import static java.util.stream.Collectors.counting;
import static java.util.stream.Collectors.groupingBy;

import java.text.Normalizer;
import java.text.Normalizer.Form;
import java.util.List;
import java.util.Map;
import java.util.Set;

import com.github.pareronia.i18n_puzzles.common.Sample;
import com.github.pareronia.i18n_puzzles.common.Samples;
import com.github.pareronia.i18n_puzzles.common.SolutionBase;
import com.github.pareronia.i18n_puzzles.common.Utils;

public class Puzzle2025_08 extends SolutionBase<Integer> {

    private Puzzle2025_08(final boolean debug) {
        super(debug);
    }

    public static Puzzle2025_08 create() {
        return new Puzzle2025_08(false);
    }

    public static Puzzle2025_08 createDebug() {
        return new Puzzle2025_08(true);
    }

    private boolean isValid(final String password) {
        final long count = password.chars().count();
        if (count < 4 || count > 12) {
            return false;
        }
        final String normalized
                = Normalizer.normalize(password.toLowerCase(), Form.NFD);
        final int bits = Utils.asCharacterStream(normalized)
                .mapToInt(ch -> {
                    int ans = 0;
                    if (Character.isDigit(ch)) {
                        ans |= 0b001;
                    }
                    if (Character.isLetter(ch)) {
                        if (Set.of('a', 'e', 'i', 'o', 'u').contains(ch)) {
                            ans |= 0b010;
                        } else {
                            ans |= 0b100;
                        }
                    }
                    return ans;
                })
                .reduce((a, b) -> a | b).getAsInt();
        final Map<Character, Long> counter = Utils.asCharacterStream(normalized)
                .filter(Character::isLetter)
                .collect(groupingBy(o -> o, counting()));
        return bits == 0b111
                && counter.values().stream().noneMatch(v -> v > 1);
    }
    
    @Override
    public Integer solve(final List<String> input) {
        return (int) input.stream().filter(this::isValid).count();
    }

    @Override
    @Samples({
        @Sample(input = TEST, expected = "2")
    })
    public void samples() { }

    public static void main(final String[] args) throws Exception {
        Puzzle2025_08.create().run();
    }

    private static final String TEST = """
            iS0
            V8AeC1S7KhP4Ļu
            pD9Ĉ*jXh
            E1-0
            ĕnz2cymE
            tqd~üō
            IgwQúPtd9
            k2lp79ąqV
            """;
}

File: src/main/python/Puzzle2025_08.py
import operator
import sys
import unicodedata
from collections import Counter
from functools import reduce

from i18n_puzzles.common import InputData
from i18n_puzzles.common import SolutionBase
from i18n_puzzles.common import i18n_puzzles_samples

TEST = """\
iS0
V8AeC1S7KhP4Ļu
pD9Ĉ*jXh
E1-0
ĕnz2cymE
tqd~üō
IgwQúPtd9
k2lp79ąqV
"""

Output = int


class Solution(SolutionBase[Output]):
    def solve(self, input: InputData) -> Output:
        def is_valid(line: str) -> bool:
            normalized = unicodedata.normalize("NFD", line.lower())
            c = Counter(ch for ch in normalized if ch.isalpha())
            return (
                4 <= len(line) <= 12
                and not any(v > 1 for v in c.values())
                and reduce(
                    operator.or_,
                    map(
                        lambda ch: ch.isnumeric() * 0b001
                        + (ch in {"a", "e", "i", "o", "u"}) * 0b010
                        + (
                            ch.isalpha()
                            and ch not in {"a", "e", "i", "o", "u"}
                        )
                        * 0b100,
                        (ch for ch in normalized),
                    ),
                )
                == 0b111
            )

        return sum(is_valid(line) for line in input)

    @i18n_puzzles_samples(((TEST, 2),))
    def samples(self) -> None:
        pass


solution = Solution(8)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()
